lib/content/write.js

```diff
@@ -1,160 +1,205 @@
-  ret.once('error', e => {
+    }
+  )
+  ret.once('error', (e) => {
     inputErr = e
   })
   return ret
 }
 
 function handleContent (inputStream, cache, opts, errCheck) {
-  return BB.using(makeTmp(cache, opts), tmp => {
+  return makeTmp(cache, opts).then((tmp) => {
     errCheck()
-    return pipeToTmp(
-      inputStream, cache, tmp.target, opts, errCheck
-    ).then(res => {
-      return moveToDestination(
-        tmp, cache, res.integrity, opts, errCheck
-      ).then(() => res)
-    })
+    return pipeToTmp(inputStream, cache, tmp.target, opts, errCheck)
+      .then((res) => {
+        return moveToDestination(
+          tmp,
+          cache,
+          res.integrity,
+          opts,
+          errCheck
+        ).then(() => res)
+      })
+      .then((result) => makeTmpDisposer(tmp, result))
+      .catch((err) => makeTmpDisposer(tmp, err, true))
   })
 }
 
 function pipeToTmp (inputStream, cache, tmpTarget, opts, errCheck) {
-  return BB.resolve().then(() => {
+  return Promise.resolve().then(() => {
     let integrity
     let size
-    const hashStream = ssri.integrityStream({
-      integrity: opts.integrity,
-      algorithms: opts.algorithms,
-      size: opts.size
-    }).on('integrity', s => {
-      integrity = s
-    }).on('size', s => {
-      size = s
-    })
+    const hashStream = ssri
+      .integrityStream({
+        integrity: opts.integrity,
+        algorithms: opts.algorithms,
+        size: opts.size
+      })
+      .on('integrity', (s) => {
+        integrity = s
+      })
+      .on('size', (s) => {
+        size = s
+      })
     const outStream = fs.createWriteStream(tmpTarget, {
       flags: 'wx'
     })
     errCheck()
-    return pipe(inputStream, hashStream, outStream).then(() => {
-      return { integrity, size }
-    }).catch(err => {
-      return rimraf(tmpTarget).then(() => { throw err })
-    })
+    return pipe(
+      inputStream,
+      hashStream,
+      outStream
+    )
+      .then(() => {
+        return { integrity, size }
+      })
+      .catch((err) => {
+        return rimraf(tmpTarget).then(() => {
+          throw err
+        })
+      })
   })
 }
 
 function makeTmp (cache, opts) {
   const tmpTarget = uniqueFilename(path.join(cache, 'tmp'), opts.tmpPrefix)
-  return fixOwner.mkdirfix(
-    cache, path.dirname(tmpTarget)
-  ).then(() => ({
+  return fixOwner.mkdirfix(cache, path.dirname(tmpTarget)).then(() => ({
     target: tmpTarget,
     moved: false
-  })).disposer(tmp => (!tmp.moved && rimraf(tmp.target)))
+  }))
+}
+
+function makeTmpDisposer (tmp, result, shouldThrow = false) {
+  const returnResult = () => {
+    if (shouldThrow) {
+      throw result
+    }
+    return result
+  }
+
+  if (tmp.moved) {
+    return returnResult()
+  }
+  return rimraf(tmp.target)
+    .then(
+      // We don't want to catch from returnResult
+      () => returnResult(),
+      // If rimraf fails, we should crash process as per bluebird
+      (err) => {
+        throw err
+      })
```

Comment:
Here is a quick example with `bluebird` it actually crashes the 
```
const Promise = require('bluebird')

function disposerFunction () {
  console.log('Running disposer')
  return Promise.reject('Disposer function rejects')
}

function resourceCreator () {
  return Promise.resolve('From Resource Creator')
    .disposer(disposerFunction)
}

Promise
  .using(resourceCreator(), (resource) => {
    console.log(resource)
    return Promise.resolve('Result of resolution doing something with resource')
  })
  .then((data) => {
    console.log('Resolved', data)
  })
  .catch((err) => {
    console.log('Rejected', err)
  })

```

```
Bills-MacBook-Pro:scratch npm$ node test.js
From Resource Creator
Running disposer

/Users/npm/Code/npm/scratch/node_modules/bluebird/js/release/using.js:12
        setTimeout(function(){throw e;}, 0);
                              ^
Disposer function success
Bills-MacBook-Pro:scratch npm$ echo $?
1
```

However, this can be caught with:
```
process.on('uncaughtException', function (err) {
  // Disposer rejected for some reason, clean up resource?
})
```

Should the utility disposer follow the bluebird approach, or should it allow rejections be caught?